File: examples/echo/echo.py
"""The reference agent: it says what you said, and sends back what you sent.

export CUCKOO_SECRET=bnd_sec_...      # from setting a socket binding
export CUCKOO_HUB=http://localhost:8080
python echo.py
"""

import logging
import os
import tempfile
from pathlib import Path

from cuckoo import Agent

logging.basicConfig(level=logging.INFO, format="%(asctime)s %(levelname)s %(message)s")
log = logging.getLogger("echo")

agent = Agent(
    secret=os.environ["CUCKOO_SECRET"],
    hub=os.environ.get("CUCKOO_HUB", "http://localhost:8080"),
)


@agent.on_message
async def handle(msg, conv):
    log.info("%s said: %s", msg.sender.display_name, msg.text or "(nothing)")

    if not msg.has_attachments:
        await conv.send(f"You said: {msg.text}")
        return

    # Files arrive as ids: the bytes are fetched only if something wants
    # them. Here everything is downloaded and sent straight back, which is
    # the smallest honest proof that both directions work.
    with tempfile.TemporaryDirectory() as tmp:
        sent = []
        for att in msg.attachments:
            path = await att.save(Path(tmp))
            log.info(
                "got %s (%s, %d bytes, %.1fs) -> %s",
                att.file_name,
                att.kind,
                att.byte_size,
                att.seconds,
                path,
            )
            # A voice note goes back as a voice note: its length and its
            # waveform travel with it, because nothing downstream can work
            # them out from the bytes.
            sent.append(
                await agent.upload(
                    path,
                    duration_ms=att.duration_ms,
                    waveform=list(att.waveform),
                    audio=att.is_audio,
                )
            )
        again = (
            "Here it is again."
            if len(sent) == 1
            else f"Here are your {len(sent)} files again."
        )
        await conv.send(again, attachments=sent)


if __name__ == "__main__":
    agent.run()

File: examples/typescript/echo.ts
/**
 * The reference agent: it says what you said.
 *
 * The same agent as `examples/echo/echo.py`, in TypeScript, so the two SDKs
 * can be checked against each other by running both against one hub.
 *
 *   export CUCKOO_SECRET=bnd_sec_...   # from setting a socket binding
 *   export CUCKOO_HUB=http://localhost:8080
 *   npm run echo
 */

import { Agent } from "@cuckoo/agent";

const secret = process.env.CUCKOO_SECRET;
if (!secret) throw new Error("set CUCKOO_SECRET to the binding secret");

const agent = new Agent({
  secret,
  hub: process.env.CUCKOO_HUB ?? "http://localhost:8080",
});

agent.onMessage(async (message, conversation) => {
  console.log(`${message.sender.displayName} said: ${message.text || "(nothing)"}`);

  if (message.attachments.length === 0) {
    await conversation.send(`You said: ${message.text}`);
    return;
  }

  // Files arrive as ids; the bytes are fetched only if something wants them.
  // Downloading and sending them straight back is the smallest honest proof
  // that both directions work.
  const sent = [];
  for (const attachment of message.attachments) {
    const bytes = await agent.client.download(attachment.mediaId);
    console.log(`got ${attachment.fileName} (${attachment.kind}, ${bytes.length} bytes)`);
    sent.push(
      await agent.client.upload(attachment.fileName || "file", bytes, {
        contentType: attachment.mimeType,
        ...(attachment.durationMs ? { durationMs: attachment.durationMs } : {}),
        ...(attachment.waveform.length ? { waveform: attachment.waveform } : {}),
        audio: attachment.kind === "audio",
      }),
    );
  }
  await conversation.send(message.text ? `You sent: ${message.text}` : "", { attachments: sent });
});

agent.onJoin(async (conversation) => {
  console.log(`joined ${conversation.id}`);
  await conversation.send("Hello. Say anything and I will say it back.");
});

console.log("echo: connecting…");
await agent.serve();
